### PR TITLE
Multi-dimensional array reordering and refactoring

### PR DESCRIPTION
**Pull request by xuanluo113**
Pull request overview
---------------------
This branch refactors some inefficient use of multi-dimensional array, including:
- Reordering some 2D window arrays to improve cache locality
- Change some 2D construction arrays to array of array to avoid slicing and reconstruction.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/General.cc [hunk 1]
```diff
@@ -1255,26 +1255,12 @@ Real64 BlindBeamBeamTrans(Real64 const ProfAng,        // Solar profile angle (r
 }
 
 Real64 POLYF(Real64 const X,         // Cosine of angle of incidence
```
Comment (amirroth) [inline src/EnergyPlus/General.cc]: Given that in a number of callsites `X` is 1, I wonder if it makes sense to make this function `constexpr`.

File: src/EnergyPlus/Construction.cc [hunk 2]
```diff
@@ -2000,8 +2000,28 @@ void ConstructionProps::setArraysBasedOnMaxSolidWinLayers(EnergyPlusData &state)
...
+        this->tBareSolCoef(Layer).dimension(DataSurfaces::MaxPolyCoeff, 0.0);
+        this->tBareVisCoef(Layer).dimension(DataSurfaces::MaxPolyCoeff, 0.0);
+        this->rfBareSolCoef(Layer).dimension(DataSurfaces::MaxPolyCoeff, 0.0);
+        this->rfBareVisCoef(Layer).dimension(DataSurfaces::MaxPolyCoeff, 0.0);
+        this->rbBareSolCoef(Layer).dimension(DataSurfaces::MaxPolyCoeff, 0.0);
+        this->rbBareVisCoef(Layer).dimension(DataSurfaces::MaxPolyCoeff, 0.0);
+        this->afBareSolCoef(Layer).dimension(DataSurfaces::MaxPolyCoeff, 0.0);
+        this->abBareSolCoef(Layer).dimension(DataSurfaces::MaxPolyCoeff, 0.0);
+    }
 }
```
Comment (amirroth) [inline src/EnergyPlus/Construction.cc:2052]: Probably doesn't make much of a difference given that this is done for constructions and only one time, but in general we should move away from combined allocation and initialization using `dimension` and towards separate allocation and explicit initialization in which all arrays can be initialized together in a single (vectorized) loop.

**Comment by amirroth** [comment]
This is a nice new code pattern! Thanks @xuanluo113!

File: src/EnergyPlus/HeatBalanceManager.cc [hunk 3]
```diff
@@ -7206,6 +7204,13 @@ namespace HeatBalanceManager {
                 state.dataConstruction->Construct(ConstrNum).TotGlassLayers = NGlass(IGlSys);
                 state.dataConstruction->Construct(ConstrNum).TotSolidLayers = NGlass(IGlSys);
 
+                for (int Layer = 1; Layer <= state.dataHeatBal->MaxSolidWinLayers; ++Layer) {
+                    for (int index = 1; index <= DataSurfaces::MaxPolyCoeff; ++index) {
+                        state.dataConstruction->Construct(ConstrNum).AbsBeamCoef(Layer)(index) = 0.0;
```
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceManager.cc:7209]: This syntax will take a while to sink in (for me anyway). `x(i)(j)`

File: src/EnergyPlus/HeatBalanceManager.cc [hunk 4]
```diff
@@ -6234,8 +6234,8 @@ namespace HeatBalanceManager {
             if (thisSurface.Class == DataSurfaces::SurfaceClass::Window) {
                 auto &thisConstruct(thisSurface.Construction);
                 if (!state.dataConstruction->Construct(thisConstruct).WindowTypeBSDF) {
-                    state.dataHeatBal->SurfWinFenLaySurfTempFront(1, SurfNum) = state.dataHeatBalSurf->TH(1, 1, SurfNum);
-                    state.dataHeatBal->SurfWinFenLaySurfTempBack(state.dataConstruction->Construct(thisConstruct).TotLayers, SurfNum) =
+                    state.dataHeatBal->SurfWinFenLaySurfTempFront(SurfNum, 1) = state.dataHeatBalSurf->TH(1, 1, SurfNum);
```
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceManager.cc:6237]: Interesting that these arrays (and many others) were originally (pre-c++) `x(SurfNum, LayerNum) (or whatever)`, but back in #4383, they were [reversed](https://github.com/NREL/EnergyPlus/pull/4843/files#diff-18fbb8a5b1334b62513de2441bdb4bb4ef1fbe5fbec24d5dd0effcd019dfc33dL721-R715) (find SolarShading.cc and expand it then the lines should show there) to be row-major, and now we're going back to the original order.

File: src/EnergyPlus/WindowManager.cc [hunk 5]
```diff
@@ -714,49 +714,49 @@ namespace WindowManager {
                              6,
                              1,
                              TotalIPhi,
-                             state.dataConstruction->Construct(ConstrNum).tBareSolCoef(_, IGlass));
+                             state.dataConstruction->Construct(ConstrNum).tBareSolCoef(IGlass));
```
Comment (mjwitte) [inline src/EnergyPlus/WindowManager.cc:717]: Always nice to get rid of `_ ,`

**Review by mjwitte — APPROVED** [review]
@xuanluo113 I'm ok with these changes. The swapped indexes in some of the 2d arrays is a bit scary, but I guess if any were missed there would be an array bounds error for sure in the CI tests.

Do you want to change the initialization method (per @amirroth [comment](https://github.com/NREL/EnergyPlus/pull/8692/files#r606707491) before this merges?

File: src/EnergyPlus/HeatBalanceManager.cc (hunk 3 above)
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceManager.cc:7209]: `x[i][j]` would make more sense after we change those arrays to vectors.

**Comment by xuanluo113** [comment]
@mjwitte I changed the initialization method in the construction array initialization. More on this topic, I'm switching to a branch to convert surface object fields to surface HB arrays. I found the location of initialization of `dataSurface`, `dataHeatBal`, `dataHeatBalSurf`, `dataHeatBalSurfMgr` arrays are sometimes random (some in SurfaceGeometry, some in SolarShading, some in HBManger or HBSurfaceManager), and it's a little misleading to justify where to add new surface HB arrays. I'll probably revisit the initialization issue later.

**Pull request merged**